src/crimson/osd/pg.cc

```diff
@@ -129,6 +130,24 @@ PG::PG(
 
 PG::~PG() {}
 
+void PG::check_blocklisted_watchers()
+{
+  logger().debug("{}", __func__);
+  shard_services.for_each_obc([this] (ObjectContextRef obc) {
```

Comment:
This is going to iterate over object contexts for other PGs as well.  The PG needs a set of its own objects with live watches -- there shouldn't in general be all that many.

Reply:
How about skipping a watch on PG mismatch?

Reply:
Iterating over all live object contexts on the shard to find the very small subset with a live watcher for this pg still strikes me as inefficient and complicated.

Reply:
I see. Perhaps we could flip the loops.

Reply:
Nope, we don't have a separate registry of watchers. The main difference with the classical is it uses a per-PG (not per-shard) registry of object contexts:

```cpp
void PrimaryLogPG::check_blocklisted_watchers()
{
  dout(20) << "PrimaryLogPG::check_blocklisted_watchers for pg " << get_pgid() << dendl;
  pair<hobject_t, ObjectContextRef> i;
  while (object_contexts.get_next(i.first, &i))
    check_blocklisted_obc_watchers(i.second);
}
```

Duplicating the information looks even worse. HOWEVER:

```cpp
  using lru_set_t = boost::intrusive::set<
    base_t,
    lru_set_option_t,
    boost::intrusive::key_of_value<VToKWrapped>
    >;
  lru_set_t lru_set;
```

Maybe, by changing the key derivation function, we could have OBCs sorted / prefixed by PG-derived part.

Reply:
Probably not, tbh.  We can just let each PG have its own.

Reply:
https://github.com/ceph/ceph/pull/49729